### PR TITLE
Add bicep files

### LINKED ISSUE #1897: The existing IaC code is migrated to Bicep and committed in bicep in the dev branch

### PR DESCRIPTION
**Pull request by TechPreacher**
# PR for issue #1897

## What is being addressed

This PR adds the first, working version of the bicep files without removing the existing ARM templates.

### CODE REVIEW

**Review by fbeltrao — APPROVED** [review]

**Review by fbeltrao — COMMENTED** [review]
Minor question I have is if there is a benefit in keeping modules in the 'modules' sub-folder. 
When developing, we mostly work with all files, having the extra folder, based on my experience, did not have value.

Wdyt?

File: TemplateBicep/main.bicep [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    location: location
+  }
+}
+
+module function 'modules/function.bicep' = {
+  name: 'function'
+  params: {
+    appInsightName: observability.outputs.appInsightName
+    logAnalyticsName: observability.outputs.logAnalyticsName
+    deployDevice: true
```
Comment (Mandur) [inline TemplateBicep/main.bicep:37]: Shouldn't these be parameters?

File: TemplateBicep/main.bicep [hunk 2]
```diff
@@ -0,0 +1,66 @@
+param location string = resourceGroup().location
+param uniqueSolutionPrefix string
+param discoveryZipUrl string
```
Comment (Mandur) [inline TemplateBicep/main.bicep:3]: As far as I can see discoveryZip Url should be static variable, it is tied to the version

Comment (Mandur) [reply TemplateBicep/main.bicep:3]: https://github.com/Azure/iotedge-lorawan-starterkit/blob/dev/Template/azuredeploy.json#L96

File: TemplateBicep/main.bicep [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+module observability 'modules/observability.bicep' = {
+  name: 'observability'
+  params: {
+    prefix: uniqueSolutionPrefix
+    iotHubName: iotHub.outputs.iotHubName
+    location: location
+  }
+}
+
+module discoveryService 'modules/discoveryService.bicep' = {
```
Comment (Mandur) [inline TemplateBicep/main.bicep:56]: The deployment of this service should be conditional based on param https://github.com/Azure/iotedge-lorawan-starterkit/blob/6a07c8f527e38506493f6f9ab0d5dabfe47d9ba6/Template/azuredeploy.json#L393

File: TemplateBicep/modules/observability.bicep [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  name: '${prefix}insight'
+  kind: 'web'
+  location: location
+  properties: {
+    WorkspaceResourceId: logAnalytics.id
+    Application_Type: 'web'
+  }
+}
+
+module workbook './workbook.bicep' = {
```
Comment (Mandur) [inline TemplateBicep/modules/observability.bicep:36]: Why having a subtemplate for one resource here?

File: TemplateBicep/modules/observability.bicep [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  name: 'workbook'
+  params: {
+    appInsightsName: appInsight.name
+    iothubName: iotHub.name
+    location: location
+  }
+}
+
+module azureMonitorAlerts './alerts.bicep' = {
```
Comment (Mandur) [inline TemplateBicep/modules/observability.bicep:45]: This deployment should be conditional https://github.com/Azure/iotedge-lorawan-starterkit/blob/dev/Template/azuredeploy.json#L315

File: TemplateBicep/main.bicep [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    name: '${uniqueSolutionPrefix}hub'
+    location: location
+  }
+}
+
+module storage 'modules/storage.bicep' = {
+  name: 'storage'
+  params: {
+    storageAccountName: '${uniqueSolutionPrefix}storage'
+    storageAccountType: 'StorageV2'
```
Comment (Mandur) [inline TemplateBicep/main.bicep:17]: I think these could be switched to internal variables, we don't expect users to switch them

File: TemplateBicep/main.bicep (hunk 1 above)
Comment (Mandur) [inline TemplateBicep/main.bicep:37]: deployDevice useAzureMonitorOnEdge Should be user defined params

File: TemplateBicep/modules/alerts.bicep [hunk 7]
```diff
@@ -0,0 +1,271 @@
+param iothubName string
+param appInsightsName string
+
+@description('Dynamic Threshold sensitivity (High, Medium, Low)')
+param alertSensitivity string = 'Medium'
```
Comment (Mandur) [inline TemplateBicep/modules/alerts.bicep:5]: nit: I think all below could be defined as variables as the main template has no option to change them

File: TemplateBicep/modules/function.bicep [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+@description('Hosting Plan SKU tiers. Default value is Dynamic.')
+param hostingPlanSkuTier string = 'Dynamic'
+
+@description('Prefix used for resource names. Should be unique as this will also be used for domain names.')
+param uniqueSolutionPrefix string
+
+@description('Controls whether observability is set up for IoT Edge.')
+param useAzureMonitorOnEdge bool
+
+@description('The Git version to use. Default is 2.1.0.')
```
Comment (Mandur) [inline TemplateBicep/modules/function.bicep:22]: This param should be propagated to the main. Bicep (we need to change it at every release, would be nice to have a central location)

File: TemplateBicep/modules/iothub.bicep [hunk 9]
```diff
@@ -0,0 +1,13 @@
+param name string
+param location string
+
+resource iotHub 'Microsoft.Devices/IotHubs@2021-07-02' = {
```
Comment (Mandur) [inline TemplateBicep/modules/iothub.bicep:4]: Nit. not sure we need a module for this single resource

**Review by Mandur — APPROVED** [review]
Very good work, thank you!

File: TemplateBicep/main.bicep (hunk 2 above)
Comment (fbeltrao) [reply TemplateBicep/main.bicep]: Changes added. Please review.

File: TemplateBicep/main.bicep (hunk 6 above)
Comment (fbeltrao) [reply TemplateBicep/main.bicep:54]: Changes added. Please review.

File: TemplateBicep/main.bicep (hunk 1 above)
Comment (fbeltrao) [reply TemplateBicep/main.bicep]: Changes added. Please review.

File: TemplateBicep/main.bicep (hunk 3 above)
Comment (fbeltrao) [reply TemplateBicep/main.bicep]: Changes added. Please review.

File: TemplateBicep/modules/alerts.bicep (hunk 7 above)
Comment (fbeltrao) [reply TemplateBicep/modules/alerts.bicep]: I will leave it for now. If we ever want to make them customizable, we only need to enable them on the main.bicep.

File: TemplateBicep/modules/iothub.bicep (hunk 9 above)
Comment (fbeltrao) [reply TemplateBicep/modules/iothub.bicep]: For now, I will leave each sub-resource in own file. Let's revisit once we start working more with them.

File: TemplateBicep/modules/observability.bicep (hunk 4 above)
Comment (fbeltrao) [reply TemplateBicep/modules/observability.bicep]: For now, I will leave each sub-resource in own file. Let's revisit once we start working more with them.

[the remark this answers is not in the document]
Comment (fbeltrao) [reply TemplateBicep/modules/observability.bicep]: Changes added. Please review.

File: TemplateBicep/modules/observability.bicep (hunk 5 above)
Comment (fbeltrao) [reply TemplateBicep/modules/observability.bicep]: Changes added. Please review.

File: TemplateBicep/main.bicep (hunk 1 above)
Comment (fbeltrao) [reply TemplateBicep/main.bicep]: Changes added. Please review.

File: TemplateBicep/main.bicep (hunk 6 above)
Comment (Mandur) [reply TemplateBicep/main.bicep:54]: LOGTM

**Pull request merged**